Flash: Rework STM32F3 flash driver and configure F3 based boards for tests

erwango opened the pull request:
A common driver is shared between stm32 chips, which was not the case for stm32f3 series. As a consequence stm32f3 was not maintained equally and was missing features such as flash layout or dts based configuration.

Besides, drivers had some flows such as wrong bus clock and missing HSI clock activation which lead to issues on boards not using HSI as main clock.
As a consequence this commit moves stm32f3 series flash driver to common stm32 flash drivers.

Tested with stm32f3_disco and nucleo_f334r8  on `samples/drivers/flash_shell` and `samples/subsys/nvs`.

Fixes #4197
Fixes #9686
Fixes part of STM32 related flash non covered files in #12860.

erwango committed:
samples/subsys/nvs: Fix typo in samples description.

Typo fix.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: nucleo_f091rc: Increase storage partition size

Storage partition minimum size is 6kb for successful test of
nvs sample.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: nucleo_l432kc: Fix comment on partition description

Comment for storage partition could be misleading.
Clarify this is a partition used for file system.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: stm32f3_disco: Add a storage partition

Add a 6kb "storage" partition to enable nvs samples test.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
dts/arm/st: stm32f3: Provide erase and write block size

On currently supported stm32f3 chips,
erase-block-size = 2048
write-block-size = 2
Set these property in stm32f3 series root dtsi file.

Fixes #9686

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: nucleo_f334r8: Provide flash storage partition

Add 6kb storage partition and declare nvs is supported.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

superna9999 commented on drivers/flash/Kconfig.stm32:
```diff
@@ -14,11 +14,12 @@ menuconfig SOC_FLASH_STM32
 	select FLASH_PAGE_LAYOUT if SOC_SERIES_STM32F7X
 	select FLASH_PAGE_LAYOUT if SOC_SERIES_STM32L4X
 	select FLASH_HAS_PAGE_LAYOUT if SOC_SERIES_STM32F0X
-	select FLASH_HAS_PAGE_LAYOUT if SOC_SERIES_STM32F4X
```
This change seems spurious

erwango replied:
Indeed, was not intended. Txs for catch up! Fixed

superna9999 reviewed (approved):
Except the spurious change, ok for me for the driver part

erwango committed:
drivers/flash: Move STM32F3 to generic STM32 series driver

A common driver is shared between stm32 chips, which was not the case
for stm32f3 series. As a consequence stm32f3 was not maintained
equally and was missing features such as flash layout or dts based
configuration.
Besides, drivers had some flows such as wrong bus clock and
missing HSI clock activation which lead to issues on boards not
using HSI as main clock.
As a consequence this commit moves stm32f3 series flash driver to
common stm32 flash drivers.

Fixes #4197

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

aurel32 commented on dts/arm/st/f3/stm32f3.dtsi:
```diff
@@ -40,6 +40,9 @@
 				compatible = "soc-nv-flash";
 				label = "FLASH_STM32";
+
+				write-block-size = <16>;
```
According to the datasheet, the flash should be written half-word by half-word. The write-block-size value is expressed in bytes and not bits, so it should be 2 here.

erwango replied:
Fixed, txs!

aurel32 commented on drivers/flash/CMakeLists.txt:
```diff
@@ -20,8 +20,10 @@ zephyr_library_sources_ifdef(CONFIG_SOC_FLASH_STM32
-zephyr_library_sources_ifdef(CONFIG_SOC_FLASH_STM32 flash_stm32f3x.c)
-zephyr_library_sources_ifdef(CONFIG_SOC_FLASH_STM32 flash_stm32f3x_priv.c)
+zephyr_library_sources_ifdef(CONFIG_SOC_FLASH_STM32
+	flash_stm32.c
```
Not an issue, just a remark for a future improvement: given that `flash_stm32.c` is now shared between all STM32 SoCs, it might be possible to compile this file as soon as `CONFIG_SOC_FLASH_STM32` is defined, and thus out of the `CONFIG_SOC_SERIES_STM32XXX` entries.

aurel32 reviewed (changes requested):
I do not have the hardware to test that, but the changes look good to me. There is only this small issue of the `write-block-size` value.

galak reviewed (commented):
Looks good, will merge once other people's review comments are addressed.

erwango committed:
drivers/flash: Rework CMakeLists.txt for stm32 drivers

Optimize CMake rules for STM23 series driver build.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>
